pravega/pravega
# Issue 6616: Pinned Segments my not be pinned after a container restart

**Change log description**  
Fixed a bug in how Segments are pinned via `MetadataStore::registerPinnedSegment` that was directly accessing the in-memory metadata without going through the underlying implementation of metadata store (e.g., Table Segments). Also, refactored the way the internal Segment for `ContainerEventProcessor` is created and pinned to memory.

**Purpose of the change**  
Fixes #6616.

**What the code does**  
This PR fixes 2 points: i) a problem in `MetadataStore::registerPinnedSegment`, which should not be called from outside as it was being done, and ii) refactors the way the internal Segment for `ContainerEventProcessor` is created and pinned to memory. The changes are as follows:
- The old `MetadataStore::registerPinnedSegment` was internally calling `MetadataStore::submitAssignment`, which was supposed to be only accessed via subclasses that extend `MetadataStore`, like `TableMetadataStore`. We should never do that, as this method performs a change in the in-memory metadata only, so no actual persistence of the Segment is done by invoking it. This, in addition to the fact of losing the `pinned` flag and therefore having the internal Segment for `ContainerEventProcessor` as a candidate for eviction, was the root cause for the problem we have observed.
  - Now, the new method `MetadataStore::pinSegmentToMemory` replaces `MetadataStore::registerPinnedSegment` and only pins a Segment to memory if it exists.
  - To this end, we have added a new callback in `StreamSegmentContainer::pinSegment()` to perform the actual pin to memory operation. Note that, conversely to other operations, this method directly sets the `pinned` flag to the in-memory metadata of a Segment, if it does exist. This is because the `pinned` flag is a "volatile" piece of information; i.e., it is not stored neither in metadata checkpoints nor in Table Segment metadata. For this reason, we could skip the normal workflow to update the metadata of a Segment via a `ContainerMetadata UpdateTransaction`, which would require a new Operation that could break backwards compatibility. (The alternative would be to forget about being able to pin Segments to memory altogether, at least in `ContainerEventProcessor`)

- `ContainerEventProcessor` has new method `tryCreateInternalSegment()` that does the actual call to `createStreamSegment`. This ensures that the Segment is properly persisted in metadata.
  - Once the Segment is created, we pin it to memory via `MetadataStore::pinSegmentToMemory` method.

**How to verify it**  
All tests should pass, included the modified ones for the functionality changes.
System tests are passing.

_System tests log analysis_:
Without this PR, we observed in the system test logs that `ContainerEventProcessor` could be associated with new Segments having much higher ids than the expected ones. This was visible on any system test run, as it induced multiple container restarts. With this change, we always see `ContainerEventProcessor` being associated with the right Segment id:
```
2022-03-07 18:46:01,726 77027 [core-10] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[0-3]: Closing 
2022-03-07 16:48:41,423 82259 [core-2] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[0-2]: Closing EventProcessor.
2022-03-07 16:48:41,423 82259 [core-2] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[0-3]: Closing EventProcessor.
2022-03-07 18:54:33,117 82619 [core-23] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[0-2]: Closing 
2022-03-07 18:54:33,118 82620 [core-23] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[0-3]: Closing EventProcessor.
...
2022-03-07 16:50:11,619 192944 [core-16] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[1-2]: Closing 
2022-03-07 16:50:11,619 192944 [core-16] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[1-3]: Closing 
2022-03-07 18:23:13,700 192818 [core-3] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[1-2]: Closing 
2022-03-07 18:23:13,700 192818 [core-3] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[1-3]: Closing EventProcessor.
...
2022-03-07 18:41:03,673 7130061 [core-4] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[2-2]: Closing 
2022-03-07 18:41:03,673 7130061 [core-4] INFO  i.p.s.s.c.ContainerEventProcessorImpl - EventProcessor[2-3]: Closing EventProcessor.
```


_Local issue reproduction_: I have tried the same experiment as in #6616. Now, we see that in a local Pravega deployment all the Segments are being pinned, even after multiple restarts (temporarily added new log for checking this):
```
2022-03-09 11:27:42,544 139463 [core-26] INFO  i.p.s.s.containers.MetadataStore - MetadataStore[0]: Segment 3 (_system/containers/event_processor_GC.queue.0_0) is pinned? true needs to be? false.
2022-03-09 11:27:42,621 139540 [core-14] INFO  i.p.s.s.containers.MetadataStore - MetadataStore[1]: Segment 3 (_system/containers/event_processor_GC.queue.1_1) is pinned? true needs to be? false.
2022-03-09 11:27:43,395 140314 [core-13] INFO  i.p.s.s.containers.MetadataStore - MetadataStore[3]: Segment 3 (_system/containers/event_processor_GC.queue.3_3) is pinned? true needs to be? false.
2022-03-09 11:27:43,482 140401 [core-11] INFO  i.p.s.s.containers.MetadataStore - MetadataStore[2]: Segment 3 (_system/containers/event_processor_GC.queue.2_2) is pinned? true needs to be? false.
```
